### ISSUE TITLE
OSStatus error: [-25299] Security error has occurred.

### ISSUE THREAD

**Issue by mickeysox** [aws-amplify/amplify-ios#649]
**Describe the bug**

The above error occurs if I sign in and straight away run a query.  The error occurs in the following scenarios

1. Explicit ```Amplify.Auth.signIn``` only -> no error
2. Explicit sign in ```Amplify.Auth.signIn```, directly after run a query ```Amplify.API.query``` -> error
3. Explicit sign in ```Amplify.Auth.signIn```, pause the process briefly```sleep(1)```run a query ```Amplify.API.query``` -> no error

**To Reproduce**

Followed this project setup:
https://docs.amplify.aws/lib/project-setup/create-application/q/platform/ios
Then add an API
https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/ios

The following code show's the process running with and without a pause.

```
         _ = Amplify.Auth.signOut(listener: { result in
                switch result {
                case .success():
                    let password = "Abc@123!"
                    
                    _ = Amplify.Auth.signIn(username: username, password: password, listener: { result in
                        switch result {
                        case .success(let signInResult):
                            
                                //without this OSStatus error: [-25299] Security error has occurred. is reported
                                sleep(1)
                                
                                if let id = Amplify.Auth.getCurrentUser()?.userId {
                                    print("Username is: \(id)")
                                    _ = Amplify.API.query(request: .get(User.self, byId: id) ) { event in
                                        switch event {
                                        case .success(let result):
                                            switch result {
                                            case .success(let user):
                                                guard let user = user else {
                                                    print("Unable to find user")
                                                    return
                                                }
                                                print("Succesfully retrieved user")
                                            case .failure(let error):
                                                print("Failed retrieve user: \(error)")
                                            }
                                        case .failure(let error):
                                            print("Failed to get user: \(error)")
                                        }
                                    }
                                }
                            
                        case .failure(let error):
                            print("Failed to sign in \(error)")
                        }
                    })
                case .failure(let error):
                    print("Failed to sign out \(error)")
                }
        })
```

outputs:

Without sleep:

```2020-07-16 00:30:33.434412+0100 AWSFullStackSocial[48777:1788132] [Amplify] Adding plugin: <AmplifyPlugins.AWSAPIPlugin: 0x6000029e0000>)
2020-07-16 00:30:33.434700+0100 AWSFullStackSocial[48777:1788132] [Amplify] Adding plugin: AmplifyPlugins.AWSCognitoAuthPlugin)
2020-07-16 00:30:33.434913+0100 AWSFullStackSocial[48777:1788132] [Amplify] Configuring
2020-07-16 00:30:33.435072+0100 AWSFullStackSocial[48777:1788132] [Amplify] Configuration: nil
2020-07-16 00:30:33.453190+0100 AWSFullStackSocial[48777:1788132] [awsAPIPlugin] Configure finished
Successfully configured Amplify
2020-07-16 00:30:33.517497+0100 AWSFullStackSocial[48777:1788132] [Amplify] AWSMobileClient Event listener - signedOut
2020-07-16 00:30:34.412741+0100 AWSFullStackSocial[48777:1788324] [Amplify] AWSMobileClient Event listener - signedIn
Username is: 6a91b0d8-9f79-418c-a218-499cb904d46c
2020-07-16 00:30:34.416604+0100 AWSFullStackSocial[48777:1788329] [API] Starting query D81AFF9B-517E-4ADF-9410-9D52EAF57C68
2020-07-16 00:30:34.812374+0100 AWSFullStackSocial[48777:1788354] OSStatus error: [-25299] Security error has occurred.
2020-07-16 00:30:35.290835+0100 AWSFullStackSocial[48777:1788329] [API] Starting network task for query D81AFF9B-517E-4ADF-9410-9D52EAF57C68
Succesfully retrieved user
```

With sleep:

```
2020-07-16 00:38:37.863015+0100 AWSFullStackSocial[48980:1795348] [Amplify] Adding plugin: <AmplifyPlugins.AWSAPIPlugin: 0x6000003a8750>)
2020-07-16 00:38:37.863380+0100 AWSFullStackSocial[48980:1795348] [Amplify] Adding plugin: AmplifyPlugins.AWSCognitoAuthPlugin)
2020-07-16 00:38:37.863655+0100 AWSFullStackSocial[48980:1795348] [Amplify] Configuring
2020-07-16 00:38:37.863857+0100 AWSFullStackSocial[48980:1795348] [Amplify] Configuration: nil
2020-07-16 00:38:37.881445+0100 AWSFullStackSocial[48980:1795348] [awsAPIPlugin] Configure finished
Successfully configured Amplify
2020-07-16 00:38:37.994072+0100 AWSFullStackSocial[48980:1795348] [Amplify] AWSMobileClient Event listener - signedOut
2020-07-16 00:38:38.859258+0100 AWSFullStackSocial[48980:1795547] [Amplify] AWSMobileClient Event listener - signedIn
Username is: 6a91b0d8-9f79-418c-a218-499cb904d46c
2020-07-16 00:38:39.866881+0100 AWSFullStackSocial[48980:1795548] [API] Starting query 747BC1A6-361E-42FB-9860-7861DB961C54
2020-07-16 00:38:39.874099+0100 AWSFullStackSocial[48980:1795548] [API] Starting network task for query 747BC1A6-361E-42FB-9860-7861DB961C54
Succesfully retrieved user
```

**Expected behavior**

**Screenshots**

**Environment(please complete the following information):**
- xcode 11.5
- cli 4.24.1
- pods 1.0.4
- iOS Simulator 13.5

**Additional context**

**Comment by wooj2** [comment]
Hi @mickeysox ,

Thank you so much for the detailed explanation and code sample.  

I copy and pasted your code into my workspace (with and without the sleep statement) and unfortunately, I was unable to reproduce the issue you are seeing.  I suspect that there are some differences in our configurations.

1.  I am running on iOS simulator 13.3, and xcode 11.3, although, I suspect that this difference is not going to cause an issue.
2.  In addition to adding Amplify.API, I also added Amplify.Auth and using Cognito User Pools to signin/out
3.  My amplifyconfiguration looks similar to this (with relevant pieces redacted)
```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "myapi649": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                    "region": "us-east-1",
                    "authorizationType": "API_KEY",
                    "apiKey": "<REDACTED>"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify/cli",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                        "Region": "us-east-1",
                        "AuthMode": "API_KEY",
                        "ApiKey": "<REDACTED>",
                        "ClientDatabasePrefix": "<REDACTED>"
                    }
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "us-east-1:<REDACTED>",
                            "Region": "us-east-1"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "us-east-1_<REDACTED>",
                        "AppClientId": "<REDACTED>",
                        "AppClientSecret": "<REDACTED>",
                        "Region": "us-east-1"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH"
                    }
                }
            }
        }
    }
```
4.  Instantiating amplify, with logging turned on:
```
            Amplify.Logging.logLevel = .verbose
            try Amplify.add(plugin: AWSCognitoAuthPlugin())
            try Amplify.add(plugin: AWSAPIPlugin(modelRegistration: AmplifyModels()))
            try Amplify.configure()
```
Are there any other differences that you can think of that might help us better reproduce this issue?

**Comment by mickeysox** [comment]
Hi @wooj2 

Thank you for getting back to me, our configurations are slightly different I think, in that my api is using cognito user pool, so yours 
```
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "myapi649": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                    "region": "us-east-1",
                    "authorizationType": "API_KEY",
                    "apiKey": "<REDACTED>"
                }
            }
        }
    },
```

mine
```
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "awsfullstacksocial": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                    "region": "us-east-1",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    },
```

here are my configurations files:

amplifyconfiguration.json
```
{
    "UserAgent": "aws-amplify-cli/2.0",
    "Version": "1.0",
    "api": {
        "plugins": {
            "awsAPIPlugin": {
                "awsfullstacksocial": {
                    "endpointType": "GraphQL",
                    "endpoint": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                    "region": "us-east-1",
                    "authorizationType": "AMAZON_COGNITO_USER_POOLS"
                }
            }
        }
    },
    "auth": {
        "plugins": {
            "awsCognitoAuthPlugin": {
                "UserAgent": "aws-amplify/cli",
                "Version": "0.1.0",
                "IdentityManager": {
                    "Default": {}
                },
                "CredentialsProvider": {
                    "CognitoIdentity": {
                        "Default": {
                            "PoolId": "us-east-1:b2dc9160-4428-45a7-aa2a-84d9e47af4d8",
                            "Region": "us-east-1"
                        }
                    }
                },
                "CognitoUserPool": {
                    "Default": {
                        "PoolId": "us-east-1_Dlibtj1Eb",
                        "AppClientId": "<REDACTED>",
                        "AppClientSecret": "<REDACTED>",
                        "Region": "us-east-1"
                    }
                },
                "Auth": {
                    "Default": {
                        "authenticationFlowType": "USER_SRP_AUTH"
                    }
                },
                "AppSync": {
                    "Default": {
                        "ApiUrl": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
                        "Region": "us-east-1",
                        "AuthMode": "AMAZON_COGNITO_USER_POOLS",
                        "ClientDatabasePrefix": "<REDACTED>"
                    }
                }
            }
        }
    }
}

```
awsconfiguration.json

```
{
    "UserAgent": "aws-amplify/cli",
    "Version": "0.1.0",
    "IdentityManager": {
        "Default": {}
    },
    "CredentialsProvider": {
        "CognitoIdentity": {
            "Default": {
                "PoolId": "us-east-1:b2dc9160-4428-45a7-aa2a-84d9e47af4d8",
                "Region": "us-east-1"
            }
        }
    },
    "CognitoUserPool": {
        "Default": {
            "PoolId": "<REDACTED>",
            "AppClientId": "<REDACTED>",
            "AppClientSecret": "<REDACTED>",
            "Region": "us-east-1"
        }
    },
    "Auth": {
        "Default": {
            "authenticationFlowType": "USER_SRP_AUTH"
        }
    },
    "AppSync": {
        "Default": {
            "ApiUrl": "https://<REDACTED>.appsync-api.us-east-1.amazonaws.com/graphql",
            "Region": "us-east-1",
            "AuthMode": "AMAZON_COGNITO_USER_POOLS",
            "ClientDatabasePrefix": "<REDACTED>"
        }
    }
}
```

AppDelegate

```
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        
        self.setupAWS()
        return true
    }
    
    func setupAWS() {
        Amplify.Logging.logLevel = .verbose
        do {
            try Amplify.add(plugin: AWSCognitoAuthPlugin())
            try Amplify.add(plugin: AWSAPIPlugin(modelRegistration: AmplifyModels()))
            try Amplify.configure()
            print("Successfully configured Amplify")
        } catch {
            print("Error configuring Amplify: \(error)")
        }
    }
```

Other than that, I am then running the code as above.  Are you also on the latest version of the cli / pods? I am on 4.24.1 and 1.0.4.

When I originally saw this error, it was a) after I upgraded to 4.24.1 and 1.04, but I was also testing a project, with the same name (different folders, different aws configurations), one using datastore and one using API, @drochetti suggested it could be a conflict somewhere, which I agreed.  So I created a brand new project, different folders, different names, I cleared down all AWS configurations (amplify delete, on all other projects), I also deleted the apps in the simulator, and then completed re-installed the simulator.  I have also tried running Device > Erase all Content and settings, so it doesn't feel like this is a conflict of apps?).

Above I have redacted most of the configuration data, except the pool id, and the reason for that, is if you look at this screenshot, it shows where the error occurs - I put a breakpoint in 

AWSUICKeyChainStore.m 

```
+ (NSError *)securityError:(OSStatus)status
{
    NSString *message = @"Security error has occurred.";
#if TARGET_OS_MAC && !TARGET_OS_IPHONE
    CFStringRef description = SecCopyErrorMessageString(status, NULL);
    if (description) {
        message = (__bridge_transfer NSString *)description;
    }
#endif
    NSError *error = [NSError errorWithDomain:AWSUICKeyChainStoreErrorDomain code:status userInfo:@{NSLocalizedDescriptionKey: message}];
    NSLog(@"OSStatus error: [%@] %@", @(error.code), error.localizedDescription);
    return error;
}
```

which gives me the following output when it is fired:

![image](https://user-images.githubusercontent.com/5666707/87850433-2c55a680-c8e8-11ea-85f5-acf8eae2e92b.png)

It would be great if we could get to the bottom of this, I am a bit stuck for ideas on what else to try!

Any help would be appreciated.

Thank you.

**Comment by wooj2** [comment]
Hi @mickeysox,

Thank you so much for you patience and detailed notes.

We've been able to reproduce this issue, and dug into why this is happening.

As for why this is happening, it is because there is a race condition between two threads.
Thread 1: Is started asynchronously after signing-in, in order to persist the identityId in the keychain
Thread 2: Happens when calling API.query -- it is designed to get the AWSCredentials in order to construct the request.  In doing so, it also attempts to persist the identityId in the keychain.

If thread2 is started prior to thread 1 finishing, both threads will attempt to do fetch the credentials.  Once each thread gets back the credentials, the SDK attempts to saves the (same) data in the keychain.  The thread that finishes second reports an error of "OSStatus error: [-25299]", since that data already exists in your keychain (by means of the first thread).

We thought of a variety of ways to fix this, and believe that the lowest risk approach would be to serialize write access to the keychain for `setData:forKey:genericAttribute:label:comment:error:` -- but that being said, we are going keep an eye on issues and if we see an uptick in errors, we may need to roll this back.

Thanks again for reporting this issue!

**Comment by wooj2** [comment]
I am optimistically closing this issue, but feel free to re-open if there problems with regards to this issue.

Thanks again!

**Comment by mickeysox** [comment]
Hi @wooj2 

thank you very much for investigating and finding the issue.  It makes absolute sense what is happening and your proposed solution.  Has this fix been released? Or any idea when it will be released?

**Comment by wooj2** [comment]
Hey @mickeysox ,

The fix has not been integrated into an official release yet.  The latest version is of aws-sdk-ios is 2.15.0, so we are expecting the code change to get into the next release (and then picked up by the latest version of amplify-ios). 

As far as when it will get released, I am unable to guarantee a specific date or even range as to when it will be released -- although, based on our previous velocity (of aws-sdk-ios) we had 3 releases June, and 3 releases in July, so if I were to bet: I'd suspect this change would be able to get into an official release this month.

**Comment by lawmicha** [comment]
https://github.com/aws-amplify/aws-sdk-ios/pull/2900 has been released in 2.15.1
